src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/AbstractRealtimeTestResultAction.java

```diff
@@ -135,4 +153,24 @@ static void saveBuild(Run<?, ?> build) {
         }
     }
 
+    public static final int NUMBER_OF_BUILDS_TO_SEARCH = 20;
+    public static final ImmutableSet<Result> RESULTS_OF_BUILDS_TO_CONSIDER = ImmutableSet.of(Result.SUCCESS, Result.UNSTABLE);
+
+    protected static TestResult findPreviousTestResult(Run<?, ?> b) {
```

Comment:
this should also use the pluggable storage API where available, will be much simpler code and more efficient:
```
diff --git a/src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeTestResultAction.java b/src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeTestResultAction.java
index cb0832b..3e56241 100644
--- a/src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeTestResultAction.java
+++ b/src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeTestResultAction.java
@@ -34,6 +34,7 @@ import hudson.tasks.junit.TestResultAction;
 import hudson.tasks.junit.JUnitResultArchiver;
 import hudson.tasks.junit.TestResult;

+import io.jenkins.plugins.junit.storage.TestResultImpl;
 import java.io.IOException;
 import java.util.Arrays;
 import java.util.Iterator;
@@ -106,6 +107,10 @@ public class RealtimeTestResultAction extends AbstractRealtimeTestResultAction {

     @Override
     protected TestResult findPreviousTestResult() throws IOException, InterruptedException {
+        TestResultImpl pluggableStorage = result.getPluggableStorage();
+        if (pluggableStorage != null) {
+            return pluggableStorage.getPreviousResult();
+        }
         return findPreviousTestResult(run);
     }
```

https://github.com/jenkinsci/junit-plugin/blob/master/src/main/java/io/jenkins/plugins/junit/storage/TestResultImpl.java#L109

Reply:
@timja I did as you requested and now use the pluggable storage when it exists.

However, I wanted to add some unit tests to make sure it works (I looked at what you did on the junit plugin with an H2 database), and I'm wondering if it makes sense, it looks (to me) like pluggableStorage can only be null. 

Because `result.getPluggableStorage()` simply returns the internal `impl` value that was set when the `TestResult` was created. The current `TestResult` is created from the `JUnitRealtime` `parse` method which runs:

```java
return new JUnitParser(keepLongStdio, true)
        .parseResult(glob, this.run, pipelineTestDetails, ws, null, listener);
```
And that will always create a `TestResult` with the `DirectoryScanner` and not with a `TestResultImpl`.

Reply:
Looks like the plugin needs to call parse or summary depending on the result of:
`JunitTestResultStorage storage = JunitTestResultStorage.find();`
https://github.com/jenkinsci/junit-plugin/blob/3cc961f67f1e3d6cccfd2a9974d45ae4feb447cf/src/main/java/hudson/tasks/junit/JUnitResultArchiver.java#L237

https://github.com/jenkinsci/junit-plugin/blob/3cc961f67f1e3d6cccfd2a9974d45ae4feb447cf/src/main/java/hudson/tasks/junit/JUnitResultArchiver.java#L246-L252


Should be a straight forward patch, maybe separate PR though

Reply:
Need to make sure we don't end up with duplicate results though, should be okay but need to test that